File: packages/procm-sdk/src/trace.ts
import type { ProcmClient } from "./client.js";
import type { JsonValue } from "./protocol.js";

export const TRACE_MAX_BYTES = 262_144;
export const TRACE_MIN_TTL_SECONDS = 1;
export const TRACE_MAX_TTL_SECONDS = 604_800;

export interface SaveTraceOptions {
  id?: string;
  ttlSeconds?: number;
  timeout?: number;
  signal?: AbortSignal;
}

export interface TraceEnvelope<T extends JsonValue = JsonValue> {
  version: 1;
  traceId: string;
  createdAt: number;
  roomId: string;
  memberId: string;
  processId?: string;
  data: T;
}

function randomId(): string {
  return globalThis.crypto?.randomUUID?.() ?? `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}`;
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).byteLength;
}

function validateData(data: JsonValue): void {
  let serialized: string | undefined;
  try {
    serialized = JSON.stringify(data);
  } catch {
    throw new Error("trace data must be JSON serializable");
  }
  if (serialized === undefined) throw new Error("trace data must be JSON serializable");
  if (byteLength(serialized) > TRACE_MAX_BYTES) {
    throw new Error(`trace data exceeds ${TRACE_MAX_BYTES} bytes`);
  }
}

export async function saveTrace<T extends JsonValue>(
  client: ProcmClient,
  data: T,
  options: SaveTraceOptions = {},
): Promise<string> {
  if (client.connectionState !== "open") throw new Error("procm client is not connected");
  if (options.ttlSeconds !== undefined &&
      (!Number.isInteger(options.ttlSeconds) || options.ttlSeconds < TRACE_MIN_TTL_SECONDS || options.ttlSeconds > TRACE_MAX_TTL_SECONDS)) {
    throw new Error(`trace ttlSeconds must be an integer from ${TRACE_MIN_TTL_SECONDS} to ${TRACE_MAX_TTL_SECONDS}`);
  }
  validateData(data);
  if (options.signal?.aborted) throw new DOMException("trace save aborted", "AbortError");

  const attempts = options.id === undefined ? 3 : 1;
  let lastError: unknown;
  for (let attempt = 0; attempt < attempts; attempt++) {
    const traceId = options.id ?? randomId();
    const requestId = randomId();
    let timer: ReturnType<typeof setTimeout> | undefined;
    let onAbort: (() => void) | undefined;
    try {
      const request = client.requestTraceStore(requestId, traceId, data, options.ttlSeconds);
      const guarded = new Promise<string>((resolve, reject) => {
        const finish = (done: () => void) => {
          if (timer) clearTimeout(timer);
          if (onAbort) options.signal?.removeEventListener("abort", onAbort);
          done();
        };
        request.then((id) => finish(() => resolve(id)), (error) => finish(() => reject(error)));
        onAbort = () => {
          const error = new DOMException("trace save aborted", "AbortError");
          client.cancelTraceStore(requestId, error);
        };
        options.signal?.addEventListener("abort", onAbort, { once: true });
        const timeout = options.timeout ?? 10_000;
        timer = setTimeout(() => {
          const error = new Error(`trace request timed out after ${timeout}ms`) as Error & { code?: string };
          error.code = "TRACE_REQUEST_TIMEOUT";
          client.cancelTraceStore(requestId, error);
        }, timeout);
      });
      return await guarded;
    } catch (error) {
      lastError = error;
      if ((error as { code?: string }).code !== "TRACE_STORE_CONFLICT" || options.id !== undefined) throw error;
    }
  }
  throw lastError;
}

// ---------------------------------------------------------------------------
// Reading traces back through the backend's HTTP Stream MCP (trace-get tool).
// ---------------------------------------------------------------------------

const MCP_PROTOCOL_VERSION = "2025-06-18";

export interface GetTraceOptions {
  timeout?: number;
  signal?: AbortSignal;
}

interface McpToolCallResponse {
  result?: { content?: Array<{ type: string; text?: string }> };
  error?: { message?: string };
}

const mcpSessions = new WeakMap<ProcmClient, Promise<void>>();

function mcpEndpoint(client: ProcmClient): { url: string; token?: string } {
  const { url, token } = client.connectionTarget;
  if (!url) throw new Error("procm WebSocket URL is required to resolve the MCP endpoint");
  const httpUrl = new URL(url);
  httpUrl.protocol = httpUrl.protocol === "wss:" ? "https:" : "http:";
  httpUrl.pathname = "/mcp";
  httpUrl.search = "";
  return { url: httpUrl.toString(), token };
}

async function mcpPost(
  endpoint: { url: string; token?: string },
  body: Record<string, unknown>,
  timeout: number,
  signal?: AbortSignal,
): Promise<string> {
  const controller = new AbortController();
  let timedOut = false;
  const timer = setTimeout(() => {
    timedOut = true;
    controller.abort();
  }, timeout);
  const onAbort = () => controller.abort();
  signal?.addEventListener("abort", onAbort, { once: true });
  try {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json, text/event-stream",
      "MCP-Protocol-Version": MCP_PROTOCOL_VERSION,
    };
    if (endpoint.token) headers.Authorization = `Bearer ${endpoint.token}`;
    const response = await fetch(endpoint.url, {
      method: "POST",
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    if (!response.ok) throw new Error(`MCP HTTP request failed with status ${response.status}`);
    return await response.text();
  } catch (error) {
    if (timedOut) throw new Error(`MCP request timed out after ${timeout}ms`);
    throw error;
  } finally {
    clearTimeout(timer);
    signal?.removeEventListener("abort", onAbort);
  }
}

async function mcpFetch(
  endpoint: { url: string; token?: string },
  body: Record<string, unknown>,
  timeout: number,
  signal?: AbortSignal,
): Promise<McpToolCallResponse> {
  const text = await mcpPost(endpoint, body, timeout, signal);
  // Streamable HTTP replies as SSE data lines; fall back to plain JSON.
  for (const line of text.split("\n")) {
    if (!line.startsWith("data: ")) continue;
    try {
      return JSON.parse(line.slice(6)) as McpToolCallResponse;
    } catch { /* skip malformed data line */ }
  }
  return JSON.parse(text) as McpToolCallResponse;
}

async function ensureMcpSession(client: ProcmClient, timeout: number, signal?: AbortSignal): Promise<void> {
  let session = mcpSessions.get(client);
  if (!session) {
    const endpoint = mcpEndpoint(client);
    session = (async () => {
      await mcpFetch(endpoint, {
        jsonrpc: "2.0",
        id: randomId(),
        method: "initialize",
        params: {
          protocolVersion: MCP_PROTOCOL_VERSION,
          capabilities: {},
          clientInfo: { name: "procm-sdk", version: "1.0" },
        },
      }, timeout, signal);
      // A JSON-RPC notification gets no response body, so send it raw.
      await mcpPost(endpoint, {
        jsonrpc: "2.0",
        method: "notifications/initialized",
      }, timeout, signal);
    })().catch((error: unknown) => {
      mcpSessions.delete(client);
      throw error;
    });
    mcpSessions.set(client, session);
  }
  return session;
}

// Fetch a stored trace by ID through the same procm-mcp instance the client is
// connected to. Throws when the trace is unknown, expired or evicted.
export async function getTrace(client: ProcmClient, id: string, options: GetTraceOptions = {}): Promise<TraceEnvelope> {
  if (!id || typeof id !== "string") throw new Error("trace id is required");
  const timeout = options.timeout ?? 10_000;
  await ensureMcpSession(client, timeout, options.signal);
  const response = await mcpFetch(mcpEndpoint(client), {
    jsonrpc: "2.0",
    id: randomId(),
    method: "tools/call",
    params: { name: "trace-get", arguments: { id } },
  }, timeout, options.signal);
  if (response.error) throw new Error(`MCP tools/call failed: ${response.error.message ?? "unknown MCP error"}`);
  const text = response.result?.content?.find((part) => part.type === "text")?.text;
  let payload: { ok?: boolean; trace?: TraceEnvelope; error?: string | { code?: string; message?: string } } | undefined;
  try {
    payload = text === undefined ? undefined : JSON.parse(text);
  } catch {
    throw new Error("trace-get returned a malformed payload");
  }
  if (!payload?.ok || !payload.trace) {
    const detail = payload?.error;
    throw new Error(typeof detail === "string" ? detail
      : detail && typeof detail === "object" ? `${detail.code ?? "TRACE_ERROR"}: ${detail.message ?? "no message"}`
      : `trace "${id}" was not found`);
  }
  return payload.trace;
}

File: packages/procm-sdk-dart/lib/src/trace.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'client.dart';
import 'protocol.dart';
import 'utils.dart';

const int traceMaxBytes = 262144;
const int traceMinTtlSeconds = 1;
const int traceMaxTtlSeconds = 604800;

class SaveTraceOptions {
  const SaveTraceOptions({this.id, this.ttlSeconds, this.timeout, this.signal});

  final String? id;
  final int? ttlSeconds;
  final Duration? timeout;
  final ProcmSignal? signal;
}

class TraceEnvelope<T extends Object?> {
  const TraceEnvelope({
    required this.version,
    required this.traceId,
    required this.createdAt,
    required this.roomId,
    required this.memberId,
    this.processId,
    required this.data,
  });

  factory TraceEnvelope.fromJson(Map<String, Object?> json) => TraceEnvelope<T>(
        version: (json['version'] as num).toInt(),
        traceId: json['traceId'] as String,
        createdAt: (json['createdAt'] as num).toInt(),
        roomId: json['roomId'] as String,
        memberId: json['memberId'] as String,
        processId: json['processId'] as String?,
        data: json['data'] as T,
      );

  final int version;
  final String traceId;
  final int createdAt;
  final String roomId;
  final String memberId;
  final String? processId;
  final T data;

  Map<String, Object?> toJson() => {
        'version': version,
        'traceId': traceId,
        'createdAt': createdAt,
        'roomId': roomId,
        'memberId': memberId,
        if (processId != null) 'processId': processId,
        'data': data,
      };
}

void _validateData(JsonValue data) {
  String serialized;
  try {
    serialized = jsonEncode(data);
  } catch (_) {
    throw ArgumentError('trace data must be JSON serializable');
  }
  if (utf8.encode(serialized).length > traceMaxBytes) {
    throw ArgumentError('trace data exceeds $traceMaxBytes bytes');
  }
}

/// Stores [data] in the room server's trace LRU; resolves with the traceId.
Future<String> saveTrace(
  ProcmClient client,
  JsonValue data, [
  SaveTraceOptions options = const SaveTraceOptions(),
]) async {
  if (client.connectionState != ProcmConnectionState.open) {
    throw StateError('procm client is not connected');
  }
  final ttlSeconds = options.ttlSeconds;
  if (ttlSeconds != null &&
      (!_isInt(ttlSeconds) || ttlSeconds < traceMinTtlSeconds || ttlSeconds > traceMaxTtlSeconds)) {
    throw ArgumentError(
        'trace ttlSeconds must be an integer from $traceMinTtlSeconds to $traceMaxTtlSeconds');
  }
  _validateData(data);
  if (options.signal?.aborted ?? false) throw const ProcmAbortException('trace save aborted');

  final attempts = options.id == null ? 3 : 1;
  Object? lastError;
  for (var attempt = 0; attempt < attempts; attempt++) {
    final traceId = options.id ?? randomId();
    final requestId = randomId();
    final timeout = options.timeout ?? const Duration(seconds: 10);
    try {
      final request = client.requestTraceStore(requestId, traceId, data, ttlSeconds);
      final guarded = _withGuards(request, client, requestId, timeout, options.signal);
      return await guarded;
    } catch (error) {
      lastError = error;
      if (error is! ProcmServerError || error.code != 'TRACE_STORE_CONFLICT' || options.id != null) {
        rethrow;
      }
    }
  }
  throw lastError!;
}

bool _isInt(num value) => value is int || value == value.roundToDouble();

Future<String> _withGuards(
  Future<String> request,
  ProcmClient client,
  String requestId,
  Duration timeout,
  ProcmSignal? signal,
) {
  final completer = Completer<String>();
  Timer? timer;
  void Function()? removeAbort;
  void finish() {
    timer?.cancel();
    removeAbort?.call();
  }

  request.then(
    (id) {
      finish();
      if (!completer.isCompleted) completer.complete(id);
    },
    onError: (Object error) {
      finish();
      if (!completer.isCompleted) completer.completeError(error);
    },
  );
  if (signal != null) {
    removeAbort = signal.onAbort(() {
      client.cancelTraceStore(requestId, const ProcmAbortException('trace save aborted'));
    });
  }
  timer = Timer(timeout, () {
    client.cancelTraceStore(
      requestId,
      const ProcmTraceTimeout('TRACE_REQUEST_TIMEOUT', 'trace request timed out'),
    );
  });
  return completer.future;
}

/// Distinguishes saveTrace timeouts (code `TRACE_REQUEST_TIMEOUT`) for retry
/// decisions; mirrors the TS SDK's `error.code` tagging.
class ProcmTraceTimeout implements Exception {
  const ProcmTraceTimeout(this.code, this.message);

  final String code;
  final String message;

  @override
  String toString() => '$code: $message';
}

// ---------------------------------------------------------------------------
// Reading traces back through the backend's HTTP Stream MCP (trace-get tool).
// ---------------------------------------------------------------------------

const String _mcpProtocolVersion = '2025-06-18';

class GetTraceOptions {
  const GetTraceOptions({this.timeout, this.signal});

  final Duration? timeout;
  final ProcmSignal? signal;
}

({String url, String? token}) _mcpEndpoint(ProcmClient client) {
  final (:url, :token) = client.connectionTarget;
  if (url.isEmpty) {
    throw StateError('procm WebSocket URL is required to resolve the MCP endpoint');
  }
  final httpUrl = Uri.parse(url);
  return (
    url: httpUrl
        .replace(
          scheme: httpUrl.scheme == 'wss' ? 'https' : 'http',
          path: '/mcp',
          query: '',
        )
        .toString(),
    token: token,
  );
}

Future<String> _mcpPost(
  ({String url, String? token}) endpoint,
  Map<String, Object?> body,
  Duration timeout,
  ProcmSignal? signal,
) async {
  final headers = <String, String>{
    'Content-Type': 'application/json',
    'Accept': 'application/json, text/event-stream',
    'MCP-Protocol-Version': _mcpProtocolVersion,
    if (endpoint.token != null) 'Authorization': 'Bearer ${endpoint.token}',
  };
  try {
    final response = await http
        .post(Uri.parse(endpoint.url), headers: headers, body: jsonEncode(body))
        .timeout(timeout);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw StateError('MCP HTTP request failed with status ${response.statusCode}');
    }
    return utf8.decode(response.bodyBytes);
  } on TimeoutException {
    throw StateError('MCP request timed out after ${timeout.inMilliseconds}ms');
  }
}

Future<Map<String, Object?>?> _mcpFetch(
  ({String url, String? token}) endpoint,
  Map<String, Object?> body,
  Duration timeout,
  ProcmSignal? signal,
) async {
  final text = await _mcpPost(endpoint, body, timeout, signal);
  // Streamable HTTP replies as SSE data lines; fall back to plain JSON.
  for (final line in text.split('\n')) {
    if (!line.startsWith('data: ')) continue;
    try {
      final value = jsonDecode(line.substring(6));
      if (value is Map<String, Object?>) return value;
    } catch (_) {
      // skip malformed data line
    }
  }
  return jsonDecode(text) is Map<String, Object?> ? jsonDecode(text) as Map<String, Object?> : null;
}

final Expando<Future<void>> _mcpSessions = Expando<Future<void>>();

Future<void> _ensureMcpSession(
  ProcmClient client,
  Duration timeout,
  ProcmSignal? signal,
) {
  var session = _mcpSessions[client];
  if (session != null) return session;
  final endpoint = _mcpEndpoint(client);
  final completer = Completer<void>();
  session = completer.future;
  _mcpSessions[client] = session;
  () async {
    await _mcpFetch(endpoint, {
      'jsonrpc': '2.0',
      'id': randomId(),
      'method': 'initialize',
      'params': {
        'protocolVersion': _mcpProtocolVersion,
        'capabilities': <String, Object?>{},
        'clientInfo': {'name': 'procm-sdk-dart', 'version': '1.0'},
      },
    }, timeout, signal);
    // A JSON-RPC notification gets no response body, so send it raw.
    await _mcpPost(endpoint, {
      'jsonrpc': '2.0',
      'method': 'notifications/initialized',
    }, timeout, signal);
  }().then(
    (_) => completer.complete(),
    onError: (Object error) {
      _mcpSessions[client] = null;
      completer.completeError(error is Error ? error : StateError(error.toString()));
    },
  );
  return session;
}

/// Fetches a stored trace by ID through the same procm-mcp instance the client
/// is connected to. Throws when the trace is unknown, expired or evicted.
Future<TraceEnvelope<JsonValue>> getTrace(
  ProcmClient client,
  String id, [
  GetTraceOptions options = const GetTraceOptions(),
]) async {
  if (id.isEmpty) throw ArgumentError('trace id is required');
  final timeout = options.timeout ?? const Duration(seconds: 10);
  await _ensureMcpSession(client, timeout, options.signal);
  final response = await _mcpFetch(_mcpEndpoint(client), {
    'jsonrpc': '2.0',
    'id': randomId(),
    'method': 'tools/call',
    'params': {
      'name': 'trace-get',
      'arguments': {'id': id},
    },
  }, timeout, options.signal);
  final error = response?['error'];
  if (error is Map) {
    throw StateError(
        'MCP tools/call failed: ${error['message'] ?? 'unknown MCP error'}');
  }
  final result = response?['result'];
  final content = result is Map ? result['content'] : null;
  final parts = content is List ? content.whereType<Map>() : const <Map>[];
  final textPart = parts
      .cast<Map<Object?, Object?>>()
      .firstWhere((part) => part['type'] == 'text' && part['text'] is String,
          orElse: () => <Object?, Object?>{});
  final text = textPart['text'] as String?;
  Map<String, Object?>? payload;
  if (text != null) {
    try {
      final decoded = jsonDecode(text);
      payload = decoded is Map<String, Object?> ? decoded : null;
    } catch (_) {
      throw StateError('trace-get returned a malformed payload');
    }
  }
  final trace = payload?['trace'];
  if (payload?['ok'] != true || trace is! Map) {
    final detail = payload?['error'];
    if (detail is String) throw StateError(detail);
    if (detail is Map) {
      final code = detail['code'] ?? 'TRACE_ERROR';
      throw StateError('$code: ${detail['message'] ?? 'no message'}');
    }
    throw StateError('trace "$id" was not found');
  }
  return TraceEnvelope.fromJson(trace.cast<String, Object?>());
}
